### PR TITLE
[BUZZOK-30289] Add NAT middleware to execute moderations

### PR DESCRIPTION
**Pull request by jpclemens0**
# RATIONALE
NAT middleware is well suited for the pre/post-invoke structure of moderations.

## CHANGES
Add NAT middleware that applies datarobot moderations.
## Checklist
- [ ] Reviewed [guideline](https://datarobot.atlassian.net/wiki/spaces/BUZOK/pages/7305920528/REVIEW+BEFORE+COMMIT+Working+with+agentic+starter+application+and+its+components)
- [ ] Updated Changelog
- [ ] New tools in `src/datarobot_genai/drtools/` require to add `integration` tests in `tests/drmcp/integration/` and `acceptance` tests in `tests/drmcp/acceptance/` (this is a MUST)

---

> [!NOTE]
> **Medium Risk**
> Adds a large new NAT middleware that can block/replace prompts and responses and modifies streamed/non-streamed outputs to attach moderation metadata; this changes core request/response behavior and adds new heavy dependencies (including OpenTelemetry version overrides).
> 
> **Overview**
> Adds a new NAT middleware (`datarobot_moderation`) that runs DataRobot guardrails in *prescore/postscore* for both invoke and streaming paths, including prompt blocking/replacement and attaching serialized moderation metrics to outputs.
> 
> Extends `DRAgentEventResponse` with `datarobot_moderations` and updates `aggregate_dragent_event_responses` to preserve the last moderation payload while also merging `usage_metrics` and propagating the latest `model`/`original_chunk`. E2E workflows now enable the middleware across agents and assert moderation metadata is present in responses/stream chunks.
> 
> Updates packaging to ship the middleware: bumps version to `0.15.42`, adds `datarobot-moderations` to the `dragent` extra, and pins OpenTelemetry via `uv` overrides to keep optional extras resolvable; CI/test configs are adjusted accordingly.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5cae8ce49f63acf38cfe9cb727e37e65f6e3363a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/datarobot_genai/nat/datarobot_moderation_middleware.py [hunk 1]
```diff
@@ -0,0 +1,1100 @@
...
+        prompt_for_postscore: str,
+    ) -> tuple[DRAgentEventResponse | None, bool]:
+        """Return ``(replacement_response, halt_stream)``; replacement None means pass-through."""
+        if self._pipeline is None:
+            return None, False
+        if not response.events or not _events_include_text_delta_for_moderation(response.events):
+            return None, False
+        completion = dragent_event_response_to_chat_completion(response)
+        if isinstance(completion, ChatCompletionChunk):
+            return None, False
```
[the remark this answers is not in the document]
Comment (jpclemens0) [reply src/datarobot_genai/nat/datarobot_moderation_middleware.py]: We will need to add streaming again so this is will be resolved

File: src/datarobot_genai/nat/datarobot_moderation_middleware.py [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+                    buffered = item_typed
+                    continue
+                yield _enrich_response_with_moderation(item_typed, postscore_eval=postscore_eval)
+
+            # Flush after normal stream end; never yield in `finally`
+            # (illegal under GeneratorExit / PEP 525).
+            if buffered is not None:
+                yield _enrich_response_with_moderation(buffered)
+        finally:
+            _moderation_request_ctx.reset(token)
```
[the remark this answers is not in the document]
Comment (jpclemens0) [reply src/datarobot_genai/nat/datarobot_moderation_middleware.py]: The NAT middleware implementation takes care of checking self.enabled.

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -8,6 +8,7 @@ on:
 
 env:
   TASK_VERSION: 3.48.0
+  TARGET_NAME: response
```
[the remark this answers is not in the document]
Comment (jpclemens0) [reply .github/workflows/e2e.yml:11]: This is needed for loading the moderation pipeline.

**Review by mdambski — APPROVED** [review]
Generally looks good, I have one note regarding the A2A support.

Current implementation will work, but based on the review by claude opus, we will be loosing moderations metadata in the A2A paths. We will have only text and all the moderation metadata (scores etc) will get discarded when converted to text using the registered converter `convert_dragent_event_response_to_str`.

Since this PR is already big and will work (message will get blocked) I am approving now and maybe we can address this in a follow up.

Below details on this from claude:


> **Moderation metadata is lost in the A2A path**
> 
> The middleware works correctly for enforcement (block/replace), but datarobot_moderations metadata (guard scores, token counts, etc.) doesn't reach A2A clients. The issue is in NATWorkflowAgentExecutor.execute() which calls runner.result(to_type=str) — this discards everything except the text.
> 
> Proposed Fix: Override execute() in _PerUserCompatibleAgentExecutor to get the result as DRAgentEventResponse, then attach datarobot_moderations to the A2A Message.metadata field keyed by an extension URI (e.g. urn:datarobot:moderation:v1). Also advertise the extension in the agent card via AgentCapabilities.extensions. This is the A2A-spec-compliant way to carry domain-specific structured data — clients that don't understand the extension simply ignore it.

**Comment by jpclemens0** [comment]
@mdambski What do we need to change to get it work with A2A. It sounds like the converter is not sufficient.

File: e2e-tests/dragent/moderation_config.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (reza-mahdiani-m) [inline e2e-tests/dragent/moderation_config.yaml:1]: There is need to deploymend_id or llm gateway for the moderation to work well.
without them it should still work but will use defauls

File: e2e-tests/dragent/moderation_config.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    type: ootb
+  - additional_guard_config: null
+    aws_model: null
+    aws_region: null
+    description: Track the number of tokens associated with the output from the LLM,
+      and/or retrieved text from the vector database.
+    google_model: null
+    google_region: null
+    is_agentic: false
+    name: Response Tokens
```
Comment (reza-mahdiani-m) [inline e2e-tests/dragent/moderation_config.yaml:26]: There is need to define guards too. for more information about yaml file please check:
https://github.com/datarobot/moderations/blob/main/docs/GUARDRAILS.md

File: src/datarobot_genai/nat/datarobot_moderation_middleware.py [hunk 6]
```diff
@@ -0,0 +1,1756 @@
...
+    base = model_dir if model_dir is not None else os.getcwd()
+    guard_config_file = os.path.join(base, MODERATION_CONFIG_FILE_NAME)
+    if not os.path.exists(guard_config_file):
+        _logger.warning(
+            "Guard config file: %s not found; moderations will not be enforced",
+            guard_config_file,
+        )
+        return None
+
+    pipeline = ModerationPipeline.from_yaml(guard_config_file)
```
Comment (reza-mahdiani-m) [inline src/datarobot_genai/nat/datarobot_moderation_middleware.py:127]: If it is easier for you you can se from_config or from_dict

File: tests/nat/test_datarobot_moderation_middleware.py [hunk 7]
```diff
@@ -0,0 +1,1088 @@
...
+            "nested": {"x": np.float64(1.5)},
+            "ts": pd.Timestamp("2026-01-01T00:00:00Z"),
+        },
+    )
+    out = chat_completion_to_dragent_event_response(chunk)
+    assert out.datarobot_moderations is not None
+    assert out.datarobot_moderations["count"] == 42
+    assert out.datarobot_moderations["nested"]["x"] == 1.5
+    assert "2026-01-01" in str(out.datarobot_moderations["ts"])
+    out.model_dump_json()
```
Comment (reza-mahdiani-m) [inline tests/nat/test_datarobot_moderation_middleware.py:1813]: Please ensure that you have some integration tests that use real (not mocked) moderation library and ensure it works as expected

**Comment by reza-mahdiani-m** [comment]
just a few points:
as it is very big PR, ensure it covers the edge cases so it will not fail in e2e.
Moderation library needs datarobot url and token be available in environment variables.
Moderation library has some new features that may facilitate your approach they would be available today/tomorrow in Pypi.

**Review by mjnitz02 — APPROVED** [review]
Overall this LGTM. One thing maybe worth considering in the future is if the middleware concept should belong inside the `datarobot_moderations` package and just be imported. Maybe moderations should just export ag_ui compliant middleware by default instead of the `datarobot_genai` library wrapping the moderations package. This would probably make it easier to ensure that changes to `datarobot_moderations` didn't accidentally break datarobot_genai.

Just a thought.

File: tests/nat/test_datarobot_moderation_middleware.py (hunk 7 above)
Comment (jpclemens0) [reply tests/nat/test_datarobot_moderation_middleware.py:1813]: I added these.

File: e2e-tests/dragent/moderation_config.yaml (hunk 5 above)
Comment (jpclemens0) [reply e2e-tests/dragent/moderation_config.yaml:26]: I added some integration tests that can be run locally with some additional guards when DR endpoint and API token are defined. The NAT integration tests are not run on PRs though.

Comment (jpclemens0) [reply e2e-tests/dragent/moderation_config.yaml:26]: There is a conflict with `ragas` version that we need to resolve before I can expand e2e tests to use LLM gateway or other guards. I'll work on it in a followup PR.
